com.sap.cloud.lm.sl.cf.process/src/main/java/com/sap/cloud/lm/sl/cf/process/steps/BuildCloudUndeployModelStep.java

```diff
@@ -81,49 +81,71 @@ protected StepPhase executeStep(ExecutionWrapper execution) {
+        return modulesToUndeploy.stream()
+            .map(module -> module.getModuleName())
+            .noneMatch(moduleName -> existingModuleName.equals(moduleName));
+    }
+
     private void setComponentsToUndeploy(DelegateExecution context, List<String> services, List<CloudApplication> apps,
         List<ConfigurationSubscription> subscriptions) {
         StepsUtil.setSubscriptionsToDelete(context, subscriptions);
         StepsUtil.setServicesToDelete(context, services);
         StepsUtil.setAppsToUndeploy(context, apps);
     }
 
-    private List<String> computeServicesToDelete(Set<String> createdService, List<DeployedMtaModule> modulesToKeep) {
-        return createdService.stream()
-            .filter(service -> modulesToKeep.stream()
-                .noneMatch(module -> module.getServices()
-                    .contains(service)))
+    private List<String> computeServicesToDelete(List<DeployedMtaModule> modulesToKeep, Set<String> existingServices,
+        List<CloudApplicationExtended> appsToDeploy) {
+        return existingServices.stream()
+            .filter(service -> shouldDeleteService(modulesToKeep, service, appsToDeploy))
             .collect(Collectors.toList());
     }
 
+    private boolean shouldDeleteService(List<DeployedMtaModule> modulesToKeep, String service,
+        List<CloudApplicationExtended> appsToDeploy) {
+        return modulesToKeep.stream()
+            .noneMatch(module -> module.getServices()
+                .contains(service))
+                && appsToDeploy.stream()
+                    .map(CloudApplicationExtended::getServices)
+                    .noneMatch(appService -> appService.equals(service));
+    }
+
     private List<DeployedMtaModule> computeModulesToUndeploy(DeployedMta deployedMta, Set<String> mtaModules, List<String> appsToDeploy) {
         return deployedMta.getModules()
             .stream()
-            .filter(deployedModule -> {
-                if (!mtaModules.contains(deployedModule.getModuleName())) {
-                    return true;
-                }
-                // The deployed module may be in the list of MTA modules, but the actual application that was created from it may have a
-                // different name:
-                return !appsToDeploy.contains(deployedModule.getAppName());
-            })
+            .filter(deployedModule -> this.shouldUndeployModule(deployedModule, mtaModules, appsToDeploy))
             .collect(Collectors.toList());
     }
 
+    private boolean shouldUndeployModule(DeployedMtaModule deployedModule, Set<String> mtaModules, List<String> appsToDeploy) {
+        if (!mtaModules.contains(deployedModule.getModuleName())) {
+            return true;
+        }
+        // The deployed module may be in the list of MTA modules, but the actual application that was created from it may have a
+        // different name:
+        boolean appNameMatchesExisting = appsToDeploy.contains(deployedModule.getAppName());
```

Comment:
Directly return instead of doing an assignment and then using a "logical not" operator.
Delete the empty line on the bottom of the method.